enhancements/microshift/microshift-apiserver-certs.md

```diff
@@ -0,0 +1,355 @@
+---
+title: microshift-apiserver-certs
+authors:
+  - "@pacevedom"
+reviewers:
+  - "@fzdarsky, MicroShift architect"
+  - "@ggiguash, MicroShift contributor"
+  - "@stlaz, Security specialist"
+  - "@pmtk, MicroShift contributor"
+  - "@copejon, MicroShift contributor"
+  - "@deads2k, OpenShift architect"
+approvers:
+  - "@dhellmann"
+api-approvers:
+  - N/A
+creation-date: 2023-01-20
+last-updated: 2023-01-20
+tracking-link:
+  - https://issues.redhat.com/browse/USHIFT-716
+see-also:
+  - N/A
+replaces:
+  - N/A
+superseded-by:
+  - N/A
+---
+
+# MicroShift API server certificates
+## Summary
+This enhancement proposes a strategy to manage kube API server certificates
+in all environments where MicroShift might be deployed, both externally
+(accessible through kubeconfig) and internally (pods reaching the API server).
+
+> Anytime the document mentions API server, it refers to kube API server.
+
+MicroShift has a wide variety of different use cases when it comes to API
+server connectivity, we can find environments where the cluster needs to be
+reachable with the IP, with the hostname, by some other name having DNS, or
+a combination of them.
+
+## Motivation
+As of today, MicroShift does not cover all authentication requirements. It
+needs to be able to authenticate API server connections using the node IP
+address, the hostname, or any other domain name with DNS resolution. It
+shares certificate handling with OCP, which is a bit different in terms of
+deployment.
+
+Unlike OCP, MicroShift does not have a load balancer to access API server, and
+users need to be able to reach it via IP address and/or the hostname.
+Therefore, the IP and/or hostname need to be present in the certificates (CN or
+SAN). Because of how the [certificate lookup](#current-certificate-strategy)
+works for incoming connections, there can only be one certificate associated
+to an IP address. Because of MicroShift network architecture, external and
+internal clients appear to use the same server IP. Connections from both types
+of clients would use the same certificate to authenticate the connection.
+
+### User Stories
+* As a MicroShift administrator/user, I want server TLS authentication using
+the hostname.
+* As a MicroShift administrator/user, I want server TLS authentication using
+a name registered in a DNS.
+* As a MicroShift administrator/user, I want server TLS authentication using
+its IP address.
+* As a MicroShift application developer, I want server TLS authentication using
+the service name, `kubernetes`.
+* As a MicroShift application developer, I want server TLS authentication using
+the service IP.
+
+### Goals
+* Allow server TLS authentication using any combination of IP address and
+domain names.
```

Comment:
We could be explicit here about wanting to maintain as much separation as possible between the trust domains.